crates/sage/src/idpicker.rs

```diff
@@ -0,0 +1,563 @@
+                    (peptides.len(), *overall_score_dict.get(key).unwrap_or(&0)),
+                )
+            })
+            .collect();
+
+        // Sort proteins based on (peptide count, overall score)
+        let sorted_keys = score_dict
+            .into_iter()
+            .sorted_by(|a, b| b.1.cmp(&a.1))
+            .collect::<Vec<_>>();
+
+        if let Some((top_key, _)) = sorted_keys.first() {
+            // let protein = top_key[0].clone();
+            // let protein_str = format!("{}/{}", top_key.len(), top_key.iter().join("/"));
+            let arry_protein = top_key
+                .into_iter()
+                .map(|x| x.to_string())
+                .collect::<Vec<_>>();
+            // println!("{},{}", protein, protein_str);
+            // proteins.push((protein.to_string(), protein_str));
+            proteins.extend(
+                top_key
+                    .iter()
+                    .map(|x| (x.to_string(), arry_protein.clone()))
+                    .collect::<Vec<_>>(),
+            );
+
+            if let Some(peptides) = prot_dict.get(top_key).cloned() {
+                used_peptides.extend(peptides);
+            }
+        }
+    }
+
+    proteins
+}
+
+#[cfg(test)]
+mod test {
+    use crate::idpicker::get_proteingroups;
+    use std::sync::Arc;
+
+    #[test]
+    fn test_valid_proteins() {
+        let test_data = vec![
+            ("peptide_1".to_string(), "protein_7".to_string()),
+            ("peptide_2".to_string(), "protein_4".to_string()),
+            ("peptide_2".to_string(), "protein_6".to_string()),
+            ("peptide_2".to_string(), "protein_9".to_string()),
+            ("peptide_3".to_string(), "protein_1".to_string()),
+            ("peptide_4".to_string(), "protein_1".to_string()),
+            ("peptide_4".to_string(), "protein_5".to_string()),
+            ("peptide_5".to_string(), "protein_7".to_string()),
+            ("peptide_6".to_string(), "protein_3".to_string()),
+            ("peptide_6".to_string(), "protein_6".to_string()),
+            ("peptide_7".to_string(), "protein_1".to_string()),
+            ("peptide_8".to_string(), "protein_1".to_string()),
+            ("peptide_8".to_string(), "protein_2".to_string()),
+            ("peptide_8".to_string(), "protein_5".to_string()),
+            ("peptide_8".to_string(), "protein_8".to_string()),
+            ("peptide_9".to_string(), "protein_1".to_string()),
+            ("peptide_10".to_string(), "protein_4".to_string()),
+            ("peptide_10".to_string(), "protein_9".to_string()),
+        ];
+
+        let data: Vec<_> = test_data
+            .into_iter()
+            .map(|(k, v)| (Arc::new(k.to_string()), Arc::new(v.to_string())))
+            .collect();
+
+        let protein_map = get_proteingroups(data);
+
+        for (i, j) in protein_map.iter() {
+            println!("Protein  ->{:?} ,  Protein Group-> {:?} ", i, j);
+        }
+
+        assert_eq!(5, protein_map.len());
```

Comment:
Can we add an assertion for the exact mapping? Let's make sure that it is deterministic and reproducible across invocations.